x-pack/plugin/data-streams/qa/rest/src/test/java/org/elasticsearch/xpack/datastreams/DataStreamUpgradeRestIT.java

```diff
@@ -0,0 +1,231 @@
+
+        // Create a template
+        Request putComposableIndexTemplateRequest = new Request("POST", "/_index_template/mysql-error");
+        putComposableIndexTemplateRequest.setJsonEntity(
+            "{"
+                + "\"index_patterns\":[\"logs-mysql-*\"],"
+                + "\"priority\":200,"
+                + "\"composed_of\":[\"logs-mappings\",\"logs-settings\"],"
+                + "\"data_stream\":{},"
+                + "\"template\":{"
+                + "\"mappings\":{"
+                + "\"properties\":{"
+                + "\"thread\":{\"type\":\"long\"}"
+                + "}"
+                + "},"
+                + "\"settings\":{"
+                + "\"index.default_pipeline\":\"mysql-error\""
+                + "}"
+                + "}"
+                + "}"
+        );
+        assertOK(client().performRequest(putComposableIndexTemplateRequest));
+
+        // Create a data stream and index first doc
+        Request indexRequest = new Request("POST", "/logs-mysql-error/_doc");
+        indexRequest.setJsonEntity("{\"@timestamp\": \"2020-12-12\",\"message\":\"abc\",\"thread\":23}");
+        assertOK(client().performRequest(indexRequest));
+
+        // Update template
+        putComposableIndexTemplateRequest = new Request("POST", "/_index_template/mysql-error");
+        putComposableIndexTemplateRequest.setJsonEntity(
+            "{"
+                + "\"index_patterns\":[\"logs-mysql-*\"],"
+                + "\"priority\":200,"
+                + "\"composed_of\":[\"logs-mappings\",\"logs-settings\"],"
+                + "\"data_stream\":{},"
+                + "\"template\":{"
+                + "\"mappings\":{"
+                + "\"properties\":{"
+                + "\"thread\":{"
+                + "\"properties\":{"
+                + "\"id\":{\"type\":\"long\"}"
+                + "}"
+                + "}"
+                + "}"
+                + "},"
+                + "\"settings\":{"
+                + "\"index.default_pipeline\":\"mysql-error\""
+                + "}"
+                + "}"
+                + "}"
+        );
+        assertOK(client().performRequest(putComposableIndexTemplateRequest));
+
+        // Update mapping
+        Request putMappingRequest = new Request("PUT", "/logs-mysql-error/_mappings");
+        putMappingRequest.addParameters(Map.of("write_index_only", "true"));
+        putMappingRequest.setJsonEntity("{\"properties\":{\"thread\":{\"properties\":{\"id\":{\"type\":\"long\"}}}}}");
+        Exception e = expectThrows(ResponseException.class, () -> client().performRequest(putMappingRequest));
+        assertThat(e.getMessage(), containsString("can't merge a non object mapping [thread] with an object mapping"));
+
+        // Rollover
+        Request rolloverRequest = new Request("POST", "/logs-mysql-error/_rollover");
+        assertOK(client().performRequest(rolloverRequest));
+
+        // Update pipeline
```

Comment:
The update pipeline should happen before line 151, but in our case we have 2 versions of the pipeline, old and new, and the old one is remove after upgrade. Like this the old index keeps the old pipeline until the rollover happened, and the new index has the new one.

Reply:
Good to know, I will mimic that behaviour in this test as well.